File: backend/app/api/videos.py
from __future__ import annotations

from fastapi import APIRouter, File, Form, UploadFile

from app.schemas.dto import (
    RoleQueryType,
    RoleType,
    VideoActionResponse,
    VideoItem,
    VideoListResponse,
    VideoQualityPairRequest,
    VideoQualityResponse,
    VideoRenameRequest,
    VideoUploadResponse,
)
from app.services.storage import delete_video, list_videos, rename_video, save_upload
from app.services.video_quality import check_video_pair_quality, check_video_quality

router = APIRouter(prefix="/videos", tags=["videos"])


@router.post("/upload", response_model=VideoUploadResponse)
async def upload_video(
    file: UploadFile = File(...),
    role: RoleType = Form(...),
    name: str | None = Form(default=None),
):
    payload = await save_upload(file=file, role=role, name=name)
    return VideoUploadResponse(**payload)


@router.get("", response_model=VideoListResponse)
async def get_videos(role: RoleQueryType = "all"):
    items = list_videos(role=role)
    return VideoListResponse(items=items)


@router.get("/{video_id}/quality", response_model=VideoQualityResponse)
async def get_video_quality(video_id: str, role: RoleType | None = None):
    return VideoQualityResponse(**check_video_quality(video_id=video_id, role=role))


@router.post("/quality-check-pair", response_model=VideoQualityResponse)
async def check_video_quality_pair(payload: VideoQualityPairRequest):
    return VideoQualityResponse(
        **check_video_pair_quality(
            teacher_video_id=payload.teacher_video_id,
            user_video_id=payload.user_video_id,
        )
    )


@router.patch("/{video_id}", response_model=VideoActionResponse)
async def rename_video_api(video_id: str, payload: VideoRenameRequest, role: RoleType):
    item = rename_video(video_id=video_id, role=role, name=payload.name)
    return VideoActionResponse(
        ok=True,
        video_id=video_id,
        message="视频已重命名",
        item=VideoItem(**item),
    )


@router.delete("/{video_id}", response_model=VideoActionResponse)
async def delete_video_api(video_id: str, role: RoleType):
    delete_video(video_id=video_id, role=role)
    return VideoActionResponse(
        ok=True,
        video_id=video_id,
        message="视频已删除",
        item=None,
    )

File: frontend/src/api/videos.ts
import { http } from "./http";
import type {
  RoleQueryType,
  RoleType,
  VideoActionResponse,
  VideoListResponse,
  VideoQualityResponse,
  VideoUploadResponse,
} from "../types/video";

export async function uploadVideo(params: {
  file: File;
  role: RoleType;
  name?: string;
}): Promise<VideoUploadResponse> {
  const form = new FormData();
  form.append("file", params.file);
  form.append("role", params.role);
  if (params.name && params.name.trim()) {
    form.append("name", params.name.trim());
  }
  const { data } = await http.post<VideoUploadResponse>("/api/videos/upload", form, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

export async function listVideos(role: RoleQueryType = "all"): Promise<VideoListResponse> {
  const { data } = await http.get<VideoListResponse>("/api/videos", {
    params: { role },
  });
  return data;
}

export async function getVideoQuality(videoId: string, role?: RoleType): Promise<VideoQualityResponse> {
  const { data } = await http.get<VideoQualityResponse>(`/api/videos/${videoId}/quality`, {
    params: role ? { role } : undefined,
  });
  return data;
}

export async function checkVideoPairQuality(params: {
  teacherVideoId: string;
  userVideoId: string;
}): Promise<VideoQualityResponse> {
  const { data } = await http.post<VideoQualityResponse>("/api/videos/quality-check-pair", {
    teacher_video_id: params.teacherVideoId,
    user_video_id: params.userVideoId,
  });
  return data;
}

export async function renameVideo(params: {
  videoId: string;
  role: RoleType;
  name: string;
}): Promise<VideoActionResponse> {
  const { data } = await http.patch<VideoActionResponse>(`/api/videos/${params.videoId}`, { name: params.name }, {
    params: { role: params.role },
  });
  return data;
}

export async function deleteVideo(params: {
  videoId: string;
  role: RoleType;
}): Promise<VideoActionResponse> {
  const { data } = await http.delete<VideoActionResponse>(`/api/videos/${params.videoId}`, {
    params: { role: params.role },
  });
  return data;
}
